indexer/cmd/indexer.go

```diff
@@ -0,0 +1,319 @@
+				for bn := new(big.Int).Set(blockNum); bn.Cmp(endBlockNum) <= 0; bn.Add(bn, big.NewInt(1)) {
+					blockNums = append(blockNums, new(big.Int).Set(bn))
+				}
+
+				blocks, err := bi.fetchBlocks(ctx, blockNums)
+				if err != nil {
+					bi.logger.Error("failed to fetch blocks", "start", blockNum, "end", endBlockNum, "error", err)
+					continue
+				}
+
+				for _, block := range blocks {
+					bi.forwardBlockChan <- block
+					bi.lastForwardIndexedBlock.Set(block.Number())
+				}
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) fetchBackwardBlocks(ctx context.Context) {
+	ticker := time.NewTicker(bi.indexInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if bi.lastBackwardIndexedBlock.Sign() <= 0 {
+				return
+			}
+			zeroBigNum := big.NewInt(0)
+			blockNum := new(big.Int).Sub(bi.lastBackwardIndexedBlock, big.NewInt(1))
+
+			for i := 0; blockNum.Cmp(zeroBigNum) >= 0; i++ {
+				endBlockNum := new(big.Int).Sub(blockNum, big.NewInt(4))
+				if endBlockNum.Cmp(zeroBigNum) < 0 {
+					endBlockNum.Set(zeroBigNum)
+				}
+
+				blockNums := []*big.Int{}
+				for bn := new(big.Int).Set(blockNum); bn.Cmp(endBlockNum) >= 0; bn.Sub(bn, big.NewInt(1)) {
+					blockNums = append(blockNums, new(big.Int).Set(bn))
+				}
+
+				blocks, err := bi.fetchBlocks(ctx, blockNums)
+				if err != nil {
+					bi.logger.Error("failed to fetch blocks", "start", blockNum, "end", endBlockNum, "error", err)
+					break
+				}
+
+				for _, block := range blocks {
+					bi.backwardBlockChan <- block
+					bi.lastBackwardIndexedBlock.Set(block.Number())
+					if block.Number().Cmp(zeroBigNum) == 0 {
+						bi.logger.Info("done fetching backward blocks...")
+						return
+					}
+				}
+				blockNum.Sub(endBlockNum, big.NewInt(1))
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) processForwardBlocks(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-bi.forwardBlockChan:
+			if err := bi.indexBlock(ctx, block); err != nil {
+				bi.logger.Error("failed to index block", "error", err)
+			}
+			if err := bi.indexTransactions(ctx, block); err != nil {
+				bi.logger.Error("failed to index transactions", "error", err)
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) processBackwardBlocks(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-bi.backwardBlockChan:
+			if err := bi.indexBlock(ctx, block); err != nil {
+				bi.logger.Error("failed to index block", "error", err)
+			}
+			if err := bi.indexTransactions(ctx, block); err != nil {
+				bi.logger.Error("failed to index transactions", "error", err)
+			}
+			if block.Number().Cmp(big.NewInt(0)) == 0 {
+				bi.logger.Info("done processing backward blocks...")
+				return
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) indexBlock(ctx context.Context, block *types.Block) error {
+	timestamp := time.UnixMilli(int64(block.Time())).UTC().Format("2006-01-02T15:04:05.000Z")
```

Comment:
Since the time layout (`2006-01-02T15:04:05.000Z`) is used in several methods, it can be extracted to a constant.